# Assignment 3

# Nitish Kumar Reddy Vinta

In this assignment, we will focus on healthcare. This data set is made available by MIT. It contains data about 9,026 heartbeat measurements. Each row represents a single measurement (captured on a timeline). There are a total of 80 data points (columns). This is a multiclass classification task: predict whether the measurement represents a normal heartbeat or other anomalies. 

## Description of Variables

You will use the **hearbeat_cleaned.csv** data set for this assignment. Each row represents a single measurement. Columns labeled as T1 from T80 are the time steps on the timeline (there are 80 time steps, each time step has only one measurement). 

The last column is the target variable. It shows the label (category) of the measurement as follows:<br>
0 = Normal<br>
1 = Supraventricular premature beat<br>
2 = Premature ventricular contraction<br>
3 = Fusion of ventricular and normal beat<br>
4 = Unclassifiable beat

## Goal

Use the data set **hearbeat_cleaned.csv** to predict the column called **Target**. The input variables are columns labeled as **T1 to T80**. 

## Submission:

Please save and submit this Jupyter notebook file. The correctness of the code matters for your grade. **Readability and organization of your code is also important.** You may lose points for submitting unreadable/undecipherable code. Therefore, use markdown cells to create sections, and use comments where necessary.


# Note:

The data is cleaned up. There are no unqueal length sequences. And, there is no zero padding. So, you shouldn't use any `Masking` layer (like I mentioned in the lecture). 

# Read and Prepare the Data (1 points)

In [274]:
# Insert as many cells as you need for data prep

In [275]:
# Common imports
import numpy as np
import pandas as pd

In [276]:
import tensorflow as tf
from tensorflow import keras

In [277]:
import os
cwd = os.getcwd()

In [278]:
os.getcwd()

'C:\\Users\\Nitish Reddy\\Downloads'

In [279]:
data = pd.read_csv("heartbeat_cleaned.csv")

In [280]:
data.shape

(7960, 81)

In [281]:
data.head()

T1     T2     T3      T4      T5      T6      T7      T8      T9  \
0  0.987  0.892  0.461  0.1130  0.1490  0.1900  0.1650  0.1620  0.1470   
1  1.000  0.918  0.621  0.1330  0.1050  0.1250  0.1170  0.0898  0.0703   
2  1.000  0.751  0.143  0.1040  0.0961  0.0519  0.0442  0.0416  0.0364   
3  1.000  0.740  0.235  0.0464  0.0722  0.0567  0.0103  0.0155  0.0284   
4  1.000  0.833  0.309  0.0191  0.1010  0.1200  0.1040  0.0874  0.0765   

      T10  ...     T72     T73     T74     T75    T76     T77     T78    T79  \
0  0.1380  ...  0.1970  0.1970  0.1960  0.2030  0.201  0.1990  0.2010  0.205   
1  0.0781  ...  0.1950  0.1910  0.1520  0.1720  0.207  0.2110  0.2070  0.207   
2  0.0857  ...  0.2260  0.2420  0.2440  0.2860  0.468  0.8160  0.9770  0.452   
3  0.0155  ...  0.0851  0.0747  0.0515  0.0593  0.067  0.0361  0.1210  0.451   
4  0.0765  ...  0.2050  0.4210  0.8030  0.9510  0.467  0.0000  0.0519  0.082   

      T80  Target  
0  0.2080       0  
1  0.1720       0  
2  0.0519       0  
3  0.8690       0  
4  0.0628       0  

[5 rows x 81 columns]

In [282]:
y = data['Target']
x = data.drop('Target',axis=1)

In [283]:
data.head()

T1     T2     T3      T4      T5      T6      T7      T8      T9  \
0  0.987  0.892  0.461  0.1130  0.1490  0.1900  0.1650  0.1620  0.1470   
1  1.000  0.918  0.621  0.1330  0.1050  0.1250  0.1170  0.0898  0.0703   
2  1.000  0.751  0.143  0.1040  0.0961  0.0519  0.0442  0.0416  0.0364   
3  1.000  0.740  0.235  0.0464  0.0722  0.0567  0.0103  0.0155  0.0284   
4  1.000  0.833  0.309  0.0191  0.1010  0.1200  0.1040  0.0874  0.0765   

      T10  ...     T72     T73     T74     T75    T76     T77     T78    T79  \
0  0.1380  ...  0.1970  0.1970  0.1960  0.2030  0.201  0.1990  0.2010  0.205   
1  0.0781  ...  0.1950  0.1910  0.1520  0.1720  0.207  0.2110  0.2070  0.207   
2  0.0857  ...  0.2260  0.2420  0.2440  0.2860  0.468  0.8160  0.9770  0.452   
3  0.0155  ...  0.0851  0.0747  0.0515  0.0593  0.067  0.0361  0.1210  0.451   
4  0.0765  ...  0.2050  0.4210  0.8030  0.9510  0.467  0.0000  0.0519  0.082   

      T80  Target  
0  0.2080       0  
1  0.1720       0  
2  0.0519       0  
3  0.8690       0  
4  0.0628       0  

[5 rows x 81 columns]

In [284]:
#Target variables need to be an array with integer type
train_y = np.array(train_y)
test_y = np.array(test_y)

train_y = train_y.astype(np.int32)
test_y = test_y.astype(np.int32)


In [285]:
#Check the first 10 values of the train_y data set
train_y[0:10]

array([0, 0, 1, 0, 0, 0, 4, 0, 0, 0])

In [286]:
train_x= np.array(train_x)
test_x= np.array(test_x)

train_x = train_x.astype(np.float32)
test_x = test_x.astype(np.float32)

In [287]:
train_x

array([[[0.968 ],
        [0.825 ],
        [0.13  ],
        ...,
        [0.    ],
        [0.0345],
        [0.188 ]],

       [[1.    ],
        [0.922 ],
        [0.746 ],
        ...,
        [0.0216],
        [0.0302],
        [0.0474]],

       [[0.784 ],
        [0.637 ],
        [0.327 ],
        ...,
        [0.322 ],
        [0.322 ],
        [0.316 ]],

       ...,

       [[0.862 ],
        [0.941 ],
        [0.544 ],
        ...,
        [0.586 ],
        [0.568 ],
        [0.595 ]],

       [[1.    ],
        [0.905 ],
        [0.312 ],
        ...,
        [0.32  ],
        [0.319 ],
        [0.314 ]],

       [[1.    ],
        [0.735 ],
        [0.146 ],
        ...,
        [0.881 ],
        [0.84  ],
        [0.816 ]]], dtype=float32)

In [288]:
train_x = np.reshape(train_x, (train_x.shape[0], train_x.shape[1], 1))
test_x = np.reshape(test_x, (test_x.shape[0], test_x.shape[1], 1))

In [289]:
train_x.shape, train_y.shape


((5572, 80, 1), (5572,))

In [290]:
train_x

array([[[0.968 ],
        [0.825 ],
        [0.13  ],
        ...,
        [0.    ],
        [0.0345],
        [0.188 ]],

       [[1.    ],
        [0.922 ],
        [0.746 ],
        ...,
        [0.0216],
        [0.0302],
        [0.0474]],

       [[0.784 ],
        [0.637 ],
        [0.327 ],
        ...,
        [0.322 ],
        [0.322 ],
        [0.316 ]],

       ...,

       [[0.862 ],
        [0.941 ],
        [0.544 ],
        ...,
        [0.586 ],
        [0.568 ],
        [0.595 ]],

       [[1.    ],
        [0.905 ],
        [0.312 ],
        ...,
        [0.32  ],
        [0.319 ],
        [0.314 ]],

       [[1.    ],
        [0.735 ],
        [0.146 ],
        ...,
        [0.881 ],
        [0.84  ],
        [0.816 ]]], dtype=float32)

# Find the baseline (0.5 point)

In [291]:
data['Target'].value_counts()/len(data)

0    0.582035
4    0.198995
2    0.155402
1    0.055905
3    0.007663
Name: Target, dtype: float64

In [292]:
train_x.shape[1]

80

# Build a cross-sectional shallow model using Keras (with only one hidden layer) (2 points)

In [373]:
model = keras.models.Sequential([
    
    keras.layers.Flatten(input_shape=[80, 1]),
    keras.layers.Dense(5, activation='relu'),
    keras.layers.Dense(5, activation='softmax')
    
])

In [374]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = tf.keras.optimizers.Nadam(learning_rate=0.01)

# If multiclass, use "sparse_categorical_crossentropy" as the loss function
model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])


history = model.fit(train_x, train_y, epochs=25,
                    validation_data=(test_x, test_y))

Epoch 1/25
175/175 [==============================] - 1s 3ms/step - loss: 0.8711 - accuracy: 0.7148 - val_loss: 0.7795 - val_accuracy: 0.7161
Epoch 2/25
175/175 [==============================] - 0s 3ms/step - loss: 0.6618 - accuracy: 0.7872 - val_loss: 0.8024 - val_accuracy: 0.7638
Epoch 3/25
175/175 [==============================] - 0s 2ms/step - loss: 0.6102 - accuracy: 0.7922 - val_loss: 0.6318 - val_accuracy: 0.7919
Epoch 4/25
175/175 [==============================] - 0s 2ms/step - loss: 0.5795 - accuracy: 0.7979 - val_loss: 0.7028 - val_accuracy: 0.7534
Epoch 5/25
175/175 [==============================] - 0s 2ms/step - loss: 0.5637 - accuracy: 0.8067 - val_loss: 0.6435 - val_accuracy: 0.7722
Epoch 6/25
175/175 [==============================] - 0s 2ms/step - loss: 0.5535 - accuracy: 0.8160 - val_loss: 0.5734 - val_accuracy: 0.7965
Epoch 7/25
175/175 [==============================] - 0s 3ms/step - loss: 0.5467 - accuracy: 0.8193 - val_loss: 0.5676 - val_accuracy: 0.8132
Epoch 

In [375]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

[0.5755746364593506, 0.7994137406349182]

In [376]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.58
accuracy: 79.94%


In [298]:
from tensorflow.keras.callbacks import EarlyStopping


earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=1, mode='auto')

callback = [earlystop]

# Build a cross-sectional deep model using Keras (with two or more hidden layers) (2 points)

In [346]:
n_steps = 80
n_inputs = 1


model = keras.models.Sequential([
    keras.layers.SimpleRNN(3, return_sequences=True, input_shape=[n_steps, n_inputs] ),
    keras.layers.SimpleRNN(3, return_sequences=True),
     keras.layers.SimpleRNN(3, return_sequences=True),
    keras.layers.SimpleRNN(3),
    keras.layers.Dense(5, activation='softmax')
])

In [347]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=25,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/25
175/175 [==============================] - 17s 78ms/step - loss: 1.0788 - accuracy: 0.5960 - val_loss: 1.4481 - val_accuracy: 0.2693
Epoch 2/25
175/175 [==============================] - 13s 75ms/step - loss: 0.9015 - accuracy: 0.6454 - val_loss: 0.8206 - val_accuracy: 0.6826
Epoch 3/25
175/175 [==============================] - 13s 74ms/step - loss: 0.8606 - accuracy: 0.6726 - val_loss: 0.7980 - val_accuracy: 0.7006
Epoch 4/25
175/175 [==============================] - 13s 74ms/step - loss: 0.8181 - accuracy: 0.6945 - val_loss: 0.7920 - val_accuracy: 0.7031
Epoch 5/25
175/175 [==============================] - 13s 75ms/step - loss: 0.8070 - accuracy: 0.7041 - val_loss: 0.8281 - val_accuracy: 0.7127
Epoch 6/25
175/175 [==============================] - 13s 75ms/step - loss: 0.7993 - accuracy: 0.7053 - val_loss: 0.7783 - val_accuracy: 0.7140
Epoch 7/25
175/175 [==============================] - 13s 74ms/step - loss: 0.7745 - accuracy: 0.7173 - val_loss: 0.7771 - val_accuracy:

In [348]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.6659407019615173, 0.7709380388259888]

In [349]:
scores = model.evaluate(train_x, train_y, verbose=0)

scores

[0.6142394542694092, 0.7923546433448792]

In [350]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.61
accuracy: 79.24%


# Build a sequential shallow LSTM Model (with only one LSTM layer) (2 points)

In [351]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    
    keras.layers.LSTM(5, activation='softmax' , input_shape=[n_steps, n_inputs])
])

In [352]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=25,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/25
175/175 [==============================] - 8s 36ms/step - loss: 1.1801 - accuracy: 0.5808 - val_loss: 1.0853 - val_accuracy: 0.5590
Epoch 2/25
175/175 [==============================] - 6s 34ms/step - loss: 1.0650 - accuracy: 0.5870 - val_loss: 1.0584 - val_accuracy: 0.5750
Epoch 3/25
175/175 [==============================] - 6s 34ms/step - loss: 1.0535 - accuracy: 0.5867 - val_loss: 1.0732 - val_accuracy: 0.6034
Epoch 4/25
175/175 [==============================] - 6s 33ms/step - loss: 1.0444 - accuracy: 0.5957 - val_loss: 1.0444 - val_accuracy: 0.5670
Epoch 5/25
175/175 [==============================] - 6s 33ms/step - loss: 1.0391 - accuracy: 0.5917 - val_loss: 1.0675 - val_accuracy: 0.6181
Epoch 6/25
175/175 [==============================] - 6s 34ms/step - loss: 1.0353 - accuracy: 0.5953 - val_loss: 1.0366 - val_accuracy: 0.5972
Epoch 7/25
175/175 [==============================] - 6s 34ms/step - loss: 1.0304 - accuracy: 0.5975 - val_loss: 1.0329 - val_accuracy: 0.6177

In [353]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.8422425389289856, 0.6905360221862793]

In [354]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.84
accuracy: 69.05%


# Build a sequential deep LSTM Model (with only two LSTM layers) (2 points)

In [357]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.LSTM(3, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.LSTM(3, return_sequences=False),
    
    keras.layers.Dense(5, activation='softmax')
])

In [358]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=25, 
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/25
175/175 [==============================] - 19s 78ms/step - loss: 1.1574 - accuracy: 0.5958 - val_loss: 1.0793 - val_accuracy: 0.6152
Epoch 2/25
175/175 [==============================] - 12s 70ms/step - loss: 1.0474 - accuracy: 0.6289 - val_loss: 1.0558 - val_accuracy: 0.6248
Epoch 3/25
175/175 [==============================] - 12s 70ms/step - loss: 0.9961 - accuracy: 0.6441 - val_loss: 0.8941 - val_accuracy: 0.6692
Epoch 4/25
175/175 [==============================] - 12s 70ms/step - loss: 0.8601 - accuracy: 0.6814 - val_loss: 0.8866 - val_accuracy: 0.6792
Epoch 5/25
175/175 [==============================] - 12s 70ms/step - loss: 0.8092 - accuracy: 0.7109 - val_loss: 0.7964 - val_accuracy: 0.7274
Epoch 6/25
175/175 [==============================] - 12s 70ms/step - loss: 0.7495 - accuracy: 0.7408 - val_loss: 0.7437 - val_accuracy: 0.7638
Epoch 7/25
175/175 [==============================] - 12s 70ms/step - loss: 0.7084 - accuracy: 0.7687 - val_loss: 0.6837 - val_accuracy:

In [359]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.5329044461250305, 0.8345896005630493]

In [360]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.53
accuracy: 83.46%


# Build a sequential shallow GRU Model (with only one GRU layer) (2 points)

In [361]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.GRU(3, input_shape=[n_steps, n_inputs]),
    keras.layers.Dense(5, activation='softmax')
])

In [362]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=25,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/25
175/175 [==============================] - 10s 44ms/step - loss: 1.1674 - accuracy: 0.5797 - val_loss: 1.1333 - val_accuracy: 0.5800
Epoch 2/25
175/175 [==============================] - 7s 41ms/step - loss: 1.1252 - accuracy: 0.5829 - val_loss: 1.1338 - val_accuracy: 0.5800
Epoch 3/25
175/175 [==============================] - 7s 42ms/step - loss: 1.1181 - accuracy: 0.5829 - val_loss: 1.1314 - val_accuracy: 0.5800
Epoch 4/25
175/175 [==============================] - 7s 41ms/step - loss: 1.0731 - accuracy: 0.5813 - val_loss: 1.0562 - val_accuracy: 0.5616
Epoch 5/25
175/175 [==============================] - 7s 42ms/step - loss: 1.0334 - accuracy: 0.5937 - val_loss: 1.0498 - val_accuracy: 0.6101
Epoch 6/25
175/175 [==============================] - 7s 41ms/step - loss: 1.0209 - accuracy: 0.5931 - val_loss: 1.0258 - val_accuracy: 0.6034
Epoch 7/25
175/175 [==============================] - 7s 41ms/step - loss: 1.0143 - accuracy: 0.5919 - val_loss: 1.0168 - val_accuracy: 0.614

In [363]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.8743905425071716, 0.6767169237136841]

In [364]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.87
accuracy: 67.67%


# Build a sequential deep GRU Model (with only two GRU layers) (2 points)

In [367]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.GRU(7, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.GRU(6),
    keras.layers.Dense(5, activation='softmax')
])

In [368]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=25,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/25
175/175 [==============================] - 21s 91ms/step - loss: 1.1168 - accuracy: 0.5795 - val_loss: 1.0518 - val_accuracy: 0.6101
Epoch 2/25
175/175 [==============================] - 14s 83ms/step - loss: 1.0246 - accuracy: 0.6010 - val_loss: 1.0172 - val_accuracy: 0.6126
Epoch 3/25
175/175 [==============================] - 15s 84ms/step - loss: 0.9636 - accuracy: 0.6330 - val_loss: 0.9369 - val_accuracy: 0.5992
Epoch 4/25
175/175 [==============================] - 15s 83ms/step - loss: 0.6948 - accuracy: 0.7627 - val_loss: 0.5587 - val_accuracy: 0.8187
Epoch 5/25
175/175 [==============================] - 14s 82ms/step - loss: 0.5545 - accuracy: 0.8313 - val_loss: 0.5438 - val_accuracy: 0.8333
Epoch 6/25
175/175 [==============================] - 15s 83ms/step - loss: 0.4931 - accuracy: 0.8518 - val_loss: 0.4810 - val_accuracy: 0.8534
Epoch 7/25
175/175 [==============================] - 14s 82ms/step - loss: 0.4786 - accuracy: 0.8545 - val_loss: 0.4657 - val_accuracy:

In [369]:


scores = model.evaluate(test_x, test_y, verbose=0)

scores



[0.34906959533691406, 0.9036850929260254]

In [370]:
scores = model.evaluate(train_x, train_y, verbose=0)

scores

[0.3154158890247345, 0.9109835028648376]

In [371]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.32
accuracy: 91.10%


# Discussion

## List the test values of each model you built (0.5 points)

Cross-sectional shallow model using Keras (with only one hidden layer)

loss: 0.58
accuracy: 79.94%


Cross-sectional deep model using Keras (with two or more hidden layers):

loss: 0.61
accuracy: 79.24%

Sequential shallow model LSTM model ( with only one LSTM layer) :

loss: 0.84
accuracy: 69.05%

Sequential shallow model LSTM model ( with only two LSTM layer) :

loss: 0.53
accuracy: 83.46%

Shallow model using GRU model (with only one GRU layer):
    
loss: 0.87
accuracy: 67.67%

Shallow model using GRU model (with only two GRU layer) :

loss: 0.32
accuracy: 91.10%

## Which model performs the best and why? (0.5 points) 
## How does it compare to baseline? (0.5 points)



After running all the models, shallow GRU model (with only two GRU models) has the highest accuracy with 91.10%. It is my best model out of all the models.  The baseline is 58.20%, my model has highest prediction above the baseline.

This model achieved highest accuracy because I have used two hidden layer with change in nuerons. I also changed activation model to softmax, Adam as an optimizer and Sparse-categorical-crossentropy has loss funtion for my model. 